cloudevents/sdk/http_events.py

```diff
@@ -45,65 +49,110 @@ def __init__(
+            headers,
+            io.BytesIO(json.dumps(data).encode()),
             data_unmarshaller
         )
 
+        # headers validation for binary events
+        for field in event_version._ce_required_fields:
+
+            # prefixes with ce- if this is a binary event
+            fieldname = CloudEvent.field_name_modifier(field, self.isbinary)
+
+            # fields_refs holds a reference to where fields should be
+            fields_refs = headers if self.isbinary else data
+
+            fields_refs_name = 'headers' if self.isbinary else 'data'
+
+            # verify field exists else throw TypeError
+            if fieldname not in fields_refs:
+                raise TypeError(
+                    f"parameter {fields_refs_name} has no required "
+                    f"attribute {fieldname}."
+                )
+
+            elif not isinstance(fields_refs[fieldname], str):
+                raise TypeError(
+                    f"in parameter {fields_refs_name}, {fieldname} "
+                    f"expected type str but found type "
+                    f"{type(fields_refs[fieldname])}."
+                )
+
+            else:
+                self.required_attribute_values[f"ce-{field}"] = \
+                    fields_refs[fieldname]
+
+        for field in event_version._ce_optional_fields:
+            fieldname = CloudEvent.field_name_modifier(field, self.isbinary)
+            if (fieldname in fields_refs) and not \
+                    isinstance(fields_refs[fieldname], str):
+                raise TypeError(
+                    f"in parameter {fields_refs_name}, {fieldname} "
+                    f"expected type str but found type "
+                    f"{type(fields_refs[fieldname])}."
+                )
+            else:
+                self.optional_attribute_values[f"ce-{field}"] = field
+
+        # structured data is inside json resp['data']
+        self.data = copy.deepcopy(data) if self.isbinary else \
+            copy.deepcopy(data.get('data', {}))
+
+        self.headers = {
+            **self.required_attribute_values,
+            **self.optional_attribute_values
+        }
+
+    def to_request(
+        self,
+        data_unmarshaller: typing.Callable = lambda x: json.loads(
+            x.read()
+            .decode('utf-8')
+        )
+    ) -> (dict, dict):
+        converter_type = converters.TypeBinary if self.isbinary else \
+            converters.TypeStructured
+
+        headers, data = self.marshall.ToRequest(
+            self.__event,
+            converter_type,
+            data_unmarshaller
+        )
+        data = data if self.isbinary else data_unmarshaller(data)['data']
+        return headers, data
+
+    def __getitem__(self, key):
+        return self.data if key == 'data' else self.headers[key]
+
     @staticmethod
-    def is_binary_cloud_event(headers):
-        for field in base._ce_required_fields:
+    def is_binary_cloud_event(event_handler, headers):
```

Comment:
Is `event_handler` the right word? I don't quite understand it. It seems like it refers to the CloudEvent version.